todo_app/app.py

```diff
@@ -1,48 +1,121 @@
+    @login_required
+    @requires_role(['Writer', 'Admin'])
     def add_to_do():
-        repo.add_item(request.form.get('title'), request.form.get('description'), request.form.get('due-date'))
+        items_repo.add_item(request.form.get('title'), request.form.get('description'), request.form.get('due-date'))
         return redirect('/')
 
     @todo_app.route('/complete', methods=['POST'])
+    @login_required
+    @requires_role(['Writer', 'Admin'])
     def complete_to_do():
         current_item_id = request.form.get('item-id')
-        repo.update_status(current_item_id, COMPLETED)
+        items_repo.update_status(current_item_id, COMPLETED)
         return redirect('/')
 
     @todo_app.route('/uncomplete', methods=['POST'])
+    @login_required
+    @requires_role(['Writer', 'Admin'])
     def uncomplete_to_do():
         current_item_id = request.form.get('item-id')
-        repo.update_status(current_item_id, NOT_STARTED)
+        items_repo.update_status(current_item_id, NOT_STARTED)
         return redirect('/')
 
     @todo_app.route('/delete', methods=['POST'])
+    @login_required
+    @requires_role(['Writer', 'Admin'])
     def delete_to_do():
-        repo.delete_item(request.form.get('item-id'))
+        items_repo.delete_item(request.form.get('item-id'))
         return redirect('/')
 
-    return todo_app
+    @todo_app.route('/make-admin', methods=['POST'])
+    @login_required
+    @requires_role(['Admin'])
+    def make_user_admin():
+        user_repo.add_user_role(request.form.get('user-id'), 'Admin')
+        return redirect('/users')
 
+    @todo_app.route('/remove-admin', methods=['POST'])
+    @login_required
+    @requires_role(['Admin'])
+    def remove_user_admin():
+        user_repo.remove_user_role(request.form.get('user-id'), 'Admin')
+        return redirect('/users')
 
-app = create_app()
+    @todo_app.route('/make-writer', methods=['POST'])
+    @login_required
+    @requires_role(['Admin'])
+    def make_user_writer():
+        user_repo.add_user_role(request.form.get('user-id'), 'Writer')
+        return redirect('/users')
+
+    @todo_app.route('/remove-writer', methods=['POST'])
+    @login_required
+    @requires_role(['Admin'])
+    def remove_user_writer():
+        user_repo.remove_user_role(request.form.get('user-id'), 'Writer')
+        return redirect('/users')
+
+    @todo_app.route('/delete-user', methods=['POST'])
+    @login_required
+    @requires_role(['Admin'])
+    def delete_user():
+        user_repo.delete_user(request.form.get('user-id'))
+        return redirect('/users')
 
+    @todo_app.route('/login/callback')
+    def login_callback():
+        return login_callback_delegate()
 
+    def access_denied(_):
+        return render_template('forbidden.html')
 
+    todo_app.register_error_handler(401, access_denied)
+    todo_app.register_error_handler(403, access_denied)
+
+    return todo_app
+
+
+app = create_app()
+initialise_auth(app)
```

Comment:
I believe this setup breaks your production Docker container image (as the entrypoint refers to `create_app` and not the Flask object itself). This matches the 500 error you're seeing on Azure atm:
![image](https://github.com/user-attachments/assets/1649a01f-7c69-4dd1-a0cb-d60d4ac83390)

Reply:
Would be easier to move `initialize_auth` to the `create_app` function?